src/System.Private.CoreLib/shared/System/IO/Stream.cs

```diff
@@ -189,6 +189,80 @@ private int GetCopyBufferSize()
             return bufferSize;
         }
 
+        public virtual void CopyTo(ReadOnlySpanAction<byte, object?> callback, object? state, int bufferSize)
+        {
+            if (callback == null) throw new ArgumentNullException(nameof(callback));
+
+            CopyTo(new WriteCallbackStream(callback, state), bufferSize);
+        }
+
+        public virtual Task CopyToAsync(Func<ReadOnlyMemory<byte>, object?, CancellationToken, ValueTask> callback, object? state, int bufferSize, CancellationToken cancellationToken)
+        {
+            if (callback == null) throw new ArgumentNullException(nameof(callback));
+
+            return CopyToAsync(new WriteCallbackStream(callback, state), bufferSize, cancellationToken);
+        }
+
+        private sealed class WriteCallbackStream : Stream
+        {
+            private readonly ReadOnlySpanAction<byte, object?>? _action;
+            private readonly Func<ReadOnlyMemory<byte>, object?, CancellationToken, ValueTask>? _func;
+            private readonly object? _state;
+
+            public WriteCallbackStream(ReadOnlySpanAction<byte, object?> action, object? state)
+            {
+                _action = action;
+                _state = state;
+            }
+
+            public WriteCallbackStream(Func<ReadOnlyMemory<byte>, object?, CancellationToken, ValueTask> func, object? state)
+            {
+                _func = func;
+                _state = state;
+            }
+
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                Write(new ReadOnlySpan<byte>(buffer, offset, count));
+            }
+
+            public override void Write(ReadOnlySpan<byte> span)
+            {
+                if (_action != null)
+                    _action(span, _state);
+                else if (_func != null)
+                    _func(span.ToArray(), _state, CancellationToken.None).GetAwaiter().GetResult();
+                else
+                    throw new NotSupportedException();
+            }
```

Comment:
Please make this:
```C#
public override void Write(ReadOnlySpan<byte> span)
{
    if (_action != null)
    {
        _action(span, _state);
         return;
    }

    // In case a poorly implemented CopyToAsync(Stream, ...) method decides to call
    // the destination stream's Write rather than WriteAsync, we make it work, but this
    // does not need to be efficient.
    Debug.Assert(_func != null);
    _func(span.ToArray(), _state, CancellationToken.None).AsTask().GetAwaiter().GetResult();
}
```
Note the `AsTask()` in there is important.  We also do not need a fallback for if both _action and _func are null, as that can't happen.